lib/resty/wasmx/shm.lua

```diff
@@ -0,0 +1,471 @@
+        NGX_WASM_SHM_EVICTION_NONE,
+    } ngx_wa_shm_eviction_e;
+
+    typedef struct {
+        ngx_wa_shm_type_e       type;
+        ngx_wa_shm_eviction_e   eviction;
+        ngx_str_t                 name;
+        ngx_log_t                *log;
+        ngx_slab_pool_t          *shpool;
+        void                     *data;
+    } ngx_wa_shm_t;
+
+    typedef enum {
+        NGX_WA_METRIC_COUNTER,
+        NGX_WA_METRIC_GAUGE,
+        NGX_WA_METRIC_HISTOGRAM,
+    } ngx_wa_metric_type_e;
+
+
+    typedef struct {
+        ngx_uint_t                   value;
+        ngx_msec_t                   last_update;
+    } ngx_wa_metrics_gauge_t;
+
+    typedef struct {
+        uint32_t                     upper_bound;
+        uint32_t                     count;
+    } ngx_wa_metrics_bin_t;
+
+    typedef struct {
+        uint8_t                      n_bins;
+        ngx_wa_metrics_bin_t         bins[];
+    } ngx_wa_metrics_histogram_t;
+
+    typedef union {
+        ngx_uint_t                   counter;
+        ngx_wa_metrics_gauge_t       gauge;
+        ngx_wa_metrics_histogram_t  *histogram;
+    } ngx_wa_metric_val_t;
+
+    typedef struct {
+        ngx_wa_metric_type_e         metric_type;
+        ngx_wa_metric_val_t          slots[];
+    } ngx_wa_metric_t;
+
+
+    typedef void (*ngx_wa_ffi_shm_get_zones_handler_pt)(ngx_wa_shm_t *shm);
+
+
+    int ngx_wa_ffi_shm_get_zones(ngx_wa_ffi_shm_get_zones_handler_pt handler);
+    int ngx_wa_ffi_shm_get_keys(ngx_wa_shm_t *shm,
+                                ngx_uint_t n,
+                                ngx_str_t **keys);
+
+    int ngx_wa_ffi_shm_get_kv_value(ngx_wa_shm_t *shm,
+                                    ngx_str_t *k,
+                                    ngx_str_t **v,
+                                    uint32_t *cas);
+    int ngx_wa_ffi_shm_set_kv_value(ngx_wa_shm_t *shm,
+                                    ngx_str_t *k, ngx_str_t *v,
+                                    uint32_t cas, int *written);
+
+    int ngx_wa_ffi_shm_define_metric(ngx_str_t *name,
+                                     ngx_wa_metric_type_e m_type);
+    int ngx_wa_ffi_shm_record_metric(uint32_t metric_id, int value);
+    int ngx_wa_ffi_shm_increment_metric(uint32_t metric_id, int value);
+    int ngx_wa_ffi_shm_get_metric(uint32_t metric_id, ngx_str_t *name,
+                                  u_char *mbuf, size_t mbs,
+                                  u_char *hbuf, size_t hbs);
+
+    int ngx_wa_ffi_shm_one_slot_metric_size();
+    int ngx_wa_ffi_shm_max_histogram_size();
+]]
+
+
+local function get_keys(zone, max)
+    if max ~= nil and (type(max) ~= "number" or max < 1) then
+        error("n must be a positive number", 2)
+    end
+
+    local shm = zone[WASM_SHM_KEY]
+    local nkeys = C.ngx_wa_ffi_shm_get_keys(shm, 0, nil)
+    if nkeys == FFI_ERROR then
+        return nil, "failed retrieving shm zone keys total"
+    end
+
+    nkeys = (max and nkeys > max) and max or nkeys
```

Comment:
Not sure we are keeping this (why can't we just pass `max` to the FFI call?), but this looks like `math.min`.